### PR TITLE
✨clusterctl: e2e test framework

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR implements the e2e test framework providing a UX similar to the docker E2E test framework
- the test is driven by a config file so you can easily switch images/manifests from dev/ci/prod
- it is easy to switch the infrastructure provider
- it works on mac

/area clustercl
/assing @chuckha 
/assing @Arvinderpal

### CODE REVIEW

File: cmd/clusterctl/test/e2e/clusterctl.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+	yaml, err := template.Yaml()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to generate yaml for the workload cluster template")
+	}
+
+	if err := input.managementCluster.Apply(ctx, yaml); err != nil {
+		return nil, errors.Wrap(err, "failed to apply the workload cluster template")
+	}
+
+	return &WorkloadCluster{}, nil
```
Comment (Arvinderpal) [inline cmd/clusterctl/test/e2e/clusterctl.go]: Is this just a place holder? Do you intend to put anything in the WorkloadCluster struct? I believe if the workload cluster is created w/o error, then all we need is the ClusterName to interact with it. In which case, we can just remove this?

File: cmd/clusterctl/test/e2e/e2e_single_tenant_test.go [hunk 2]
```diff
@@ -18,6 +18,65 @@ limitations under the License.
...
+			// pass reference to the config for this E2E test that includes the name of the management cluster and the list of images to pre-load
+			config: config,
+		})
+		Expect(err).ToNot(HaveOccurred())
+		Expect(mgmt).ShouldNot(BeNil())
+	})
+
+	AfterEach(func() {
+		// Tears down the kind cluster
+		mgmt.Teardown(ctx, skipResourceCleanup)
```
Comment (Arvinderpal) [inline cmd/clusterctl/test/e2e/e2e_single_tenant_test.go]: Don't we need Teardown() for the workload cluster here too?

File: cmd/clusterctl/test/e2e/clusterctl.go (hunk 1 above)
Comment (fabriziopandini) [reply cmd/clusterctl/test/e2e/clusterctl.go]: now it is addressed

File: cmd/clusterctl/test/e2e/e2e_single_tenant_test.go (hunk 2 above)
Comment (fabriziopandini) [reply cmd/clusterctl/test/e2e/e2e_single_tenant_test.go]: now it is addressed

**Comment by fabriziopandini** [comment]
@vincepri @chuckha @Arvinderpal now this is ready for review.
With this new update of the clusterctl E2E test, it should be possible to use a config file to:
- switch from dev/ci/prod images
- switch the infrastructure provider from docker to *
- switch the workload cluster-template

area/clusterctl

File: cmd/clusterctl/test/README.md [hunk 3]
```diff
@@ -1,26 +1,28 @@
...
+To select a specific test configuration, use the `E2E_CONF_FILE` os environment variable:
 
-To run specific tests, use the `GINKGO_FOCUS`
+	E2E_CONF_FILE="~/my-e2e.conf" cmd/clusterctl/test/run-e2e.sh
 
-	GINKGO_FOCUS="clusterctl create cluster" ./run-e2e.sh
+To specify where test artifacts should be stored, use the `ARTIFACTS` os environment variable:
 
-## Skip local build of CAPD
+	ARTIFACTS="~/_artifacts" cmd/clusterctl/test/run-e2e.sh
```
Comment (wfernandes) [inline cmd/clusterctl/test/README.md]: Can we add this directory to the .gitignore file list?

**Comment by vincepri** [comment]
@chuckha given your experience with the other test framework, do you wanna review this?

**Comment by chuckha** [comment]
@vincepri yes, but my review is going to have to wait until the end of this week. It's too big of a PR to do right now.

That being said, I have *run* this code and it works great! My review is not necessary for a merge, so don't let me hold this PR back.

**Comment by vincepri** [comment]
good to know, thanks!

File: cmd/clusterctl/test/README.md (hunk 3 above)
Comment (fabriziopandini) [reply cmd/clusterctl/test/README.md]: I guess you mean _artifacts under the cluster-api root
done

**Comment by fabriziopandini** [comment]
rebased
@wfernandes if you follow the instruction on top of `dev-e2e.conf` and run `make -C test/infrastructure/docker docker-build REGISTRY=gcr.io/k8s-staging-capi-docker` you get the -amd64 image generated and the `test/infrastructure/docker/config/default/manager_image_patch.yaml` updated accordingly

**Comment by wfernandes** [comment]
I was able to run the e2e tests successfully with images from my own registry. 
/lgtm

A couple of thoughts:
- I suggest creating a tracking issue which includes specific e2e tests we'd like for clusterctl; similar to what we did for the capi e2e test framework. This will give other contributors a chance to add to the clusterctl e2e test suite. WDYT?
- I noticed we have a pending test regarding `clusterctl move`. Do we have a followup issue to work on that?

**Comment by fabriziopandini** [comment]
@wfernandes +1 to open an issue to follow up on this work.

**Comment by wfernandes** [comment]
/lgtm
I was able to run the tests successfully locally. Seems like there is a conflict currently.

**Comment by fabriziopandini** [comment]
/hold
as per discussion https://github.com/kubernetes-sigs/cluster-api/issues/2753#issuecomment-605893271, instead of creating a new clusterctl E2E framework we should improve the current E2E framework to support clusterctl

**Comment by vincepri** [comment]
@fabriziopandini ok to close this one?

**Pull request closed without merging**